pkg/controller/security/manager.go

```diff
@@ -0,0 +1,254 @@
+			return heap.Remove(pq, i).(*certExp)
+		}
+	}
+	return nil
+}
+
+// Find the top item and pop if it is about to expire
+func (pq *certificateQueue) lookandPop() *certExp {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	top := pq.certs[0]
+	if time.Until(top.exp.Add(-10*time.Minute)) <= 0 {
+		return heap.Pop(pq).(*certExp)
+	}
+	return nil
+}
+
+func (s *SecretManager) lookupandStoreCert(Identity string, newCert *istiosecurity.SecretItem) {
+	s.certsCache.mu.Lock()
+	defer s.certsCache.mu.Unlock()
+	// Check if the key exists in the map
+	// If refCnt == 0, then this certificate is about to be deleted, so do not perform a refresh.
+	certCache := s.certsCache.certs[Identity]
+	if certCache != nil && certCache.refCnt != 0 {
+		certCache.cert = newCert
+		certExp := certExp{exp: newCert.ExpireTime, Identity: Identity}
+		s.certsQueue.addItem(&certExp)
+		log.Infof("cert %v add, exp:%v\n", Identity, newCert.ExpireTime)
+	}
+}
+
+// Check if the certificate already exists, if it does, increment the reference count by 1,
+// if it doesn't, request a new certificate.
+func (c *certsCache) lookuporStore(Identity string) *certCache {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cert := c.certs[Identity]
+	if cert != nil {
+		cert.refCnt++
+		return cert
+	}
+	cert = &certCache{
+		refCnt: 1,
+	}
+	c.certs[Identity] = cert
+	return nil
+}
+
+// NewSecretManager creates a new secretManager.
+func NewSecretManager() (*SecretManager, error) {
+	tlsOpts = &TLSOptions{
+		RootCert: constants.RootCertPath,
+	}
+
+	options := NewSecurityOptions()
+	caClient, err := NewCaClient(options, tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+	certCache := newCertCache()
+	pq := newPriorityQueue()
+	heap.Init(pq)
+
+	secretManager := SecretManager{
+		caClient:      caClient,
+		configOptions: options,
+		certsCache:    certCache,
+		certsQueue:    pq,
+	}
+	go secretManager.refreshExpiringCerts()
+	return &secretManager, nil
+}
+
+// Automatically check and refresh when the validity period expires
+// Store the Identity in the priority queue according to the expiration time.
+// Check the highest priority element in the queue every 5 minutes.
+// If it is about to expire, pop up the element and reapply for the certificate.
+func (s *SecretManager) refreshExpiringCerts() {
+	for {
+		if s.certsQueue.Len() != 0 {
+			// As long as the memory in the Go language is referenced by any pointer,
+			// it will not be released. The map stores pointers, so it is safe here.
+			top := s.certsQueue.lookandPop()
+			if top != nil {
+				log.Debugf("refresh Identity: %v  exp: %v\n", top.Identity, top.exp)
+				newCert, err := s.caClient.fetchCert(top.Identity)
+				if err != nil {
+					log.Errorf("%v refresh fetchCert error : %v", top.Identity, err)
+					return
+				}
+				s.lookupandStoreCert(top.Identity, newCert)
+				continue
+			}
+		}
+		time.Sleep(5 * time.Minute)
+	}
+}
+
+// Initialize the certificate for the first time
+func (s *SecretManager) AddCerts(Identity string) {
```

Comment:
we should add a channel, and when `s.caClient.fetchCert(Identity)` returns false, re send to queue

Reply:
ok